go/ct/gen/interval_solver.go

```diff
@@ -0,0 +1,188 @@
+package gen
+
+import (
+	"fmt"
+	"strings"
+
+	"golang.org/x/exp/constraints"
+	"pgregory.net/rand"
+)
+
+type IntervalSolver[T constraints.Integer] struct {
+	intervals []interval[T]
+}
+
+func NewIntervalSolver[T constraints.Integer](min, max T) *IntervalSolver[T] {
+	initialInterval := interval[T]{low: min, high: max}
+	if initialInterval.isEmpty() {
+		return &IntervalSolver[T]{}
+	}
+	return &IntervalSolver[T]{
+		intervals: []interval[T]{initialInterval},
+	}
+}
+
+func (s *IntervalSolver[T]) Exclude(min, max T) {
+	if max < min {
+		return
+	}
+	res := []interval[T]{}
+	for _, current := range s.intervals {
+		newLeft := interval[T]{current.low, min - 1}
+		newRight := interval[T]{max + 1, current.high}
+
+		switch current.getRelationTo(&interval[T]{min, max}) {
+		case isEqual:
+			continue // we remove this interval
+		case isSubset:
+			res = append(res, newLeft, newRight)
+		case isSuperset:
+			continue // the current interval is dropped
+		case intersectsWithStart:
+			res = append(res, newRight)
+		case intersectsWithEnd:
+			res = append(res, newLeft)
+		default:
+			res = append(res, current)
+		}
+
+	}
+	s.intervals = res
+}
+
+func (s *IntervalSolver[T]) Contains(value T) bool {
+	for _, interval := range s.intervals {
+		if interval.contains(value) {
+			return true
+		}
+	}
+	return false
+}
+
+func (s *IntervalSolver[T]) AddEqualityConstraint(value T) {
+	s.AddLowerBoundary(value)
+	s.AddUpperBoundary(value)
+}
+
+func (s *IntervalSolver[T]) AddLowerBoundary(value T) {
+	if len(s.intervals) == 0 {
+		return
+	}
+	min := s.intervals[0].low
+	s.Exclude(min, value-1)
+}
+
+func (s *IntervalSolver[T]) AddUpperBoundary(value T) {
+	if len(s.intervals) == 0 {
+		return
+	}
+	max := s.intervals[len(s.intervals)-1].high
+	s.Exclude(value+1, max)
+}
+
+func (s *IntervalSolver[T]) IsSatisfiable() bool {
+	return len(s.intervals) > 0
+}
+
+func (s *IntervalSolver[T]) Generate(rnd *rand.Rand) (T, error) {
+	if s.intervals == nil || len(s.intervals) == 0 {
+		return 0, ErrUnsatisfiable
+	}
+
+	domainSize := uint64(0)
+	for _, interval := range s.intervals {
+		if interval.isEmpty() {
+			return 0, ErrUnsatisfiable
+		}
```

Comment:
This should not happen. There should be an invariant that makes sure that there is no empty interval in the interval list.

Reply:
The way I see it, if you have a constructor function with 2 integers, and they get misplaced, the object will be constructed in an insatisfacible state. Would it?